app/change_set_persisters/hyrax/change_set_persister.rb

```diff
@@ -3,53 +3,104 @@
+    # @return the saved resource, such as GenericWork, that is subclassed from Valkyrie::Resource
     def save(change_set:)
-      result = nil
-      catch(:abort) do
-        run_hook(:before_save, change_set: change_set)
-        result = persister.save(resource: change_set.resource)
-        run_hook(:after_save, change_set: change_set, resource: result)
-      end
-      result
+      run_hook(:before_save, change_set: change_set)
+      resource = persister.save(resource: change_set.resource)
+      run_hook(:after_save, change_set: change_set)
+      query_service.find_by(id: resource.id)
     end
 
+    # Deletes the resource contained in the change_set from the adapters specified
+    # upon initialization of the ChangeSetPersister
+    # @param change_set Any change_set created by Hyrax::DynamicChangeSet
+    # @return [Boolean] true returned if the resource in the change_set was
+    #   succesfully deleted
+    # @return [Boolean] false returned if the resource in the change_set was not deleted
     def delete(change_set:)
-      result = nil
-      catch(:abort) do
-        run_hook(:before_delete, change_set: change_set)
-        result = persister.delete(resource: change_set.resource)
-        run_hook(:after_delete, change_set: change_set)
-      end
-      result
+      run_hook(:before_delete, change_set: change_set)
+      persister.delete(resource: change_set.resource)
+      run_hook(:after_delete, change_set: change_set)
     end
 
+    # Calls save on all change_sets in the array passed in
+    # @param [Array] change_sets An array of any change_set created by Hyrax::DynamicChangeSet
+    # @return [Array] an array of the saved resources, such as GenericWork,
+    #  that is subclassed from Valkyrie::Resource
     def save_all(change_sets:)
       change_sets.map do |change_set|
         save(change_set: change_set)
       end
     end
 
-    def buffer_into_index
-      metadata_adapter.persister.buffer_into_index do |buffered_adapter|
-        with(metadata_adapter: buffered_adapter) do |buffered_changeset_persister|
-          yield(buffered_changeset_persister)
-        end
+    # Calls delete on all change_sets in the arraypassed in
+    # @param [Array] change_sets An array of any change_set created by Hyrax::DynamicChangeSet
+    # @return [Array] an array of the resources that failed to delete, if empty
+    #  all resources were succesfully deleted
+    def delete_all(change_sets:)
+      change_sets.map do |change_set|
+        delete(change_set: change_set)
       end
     end
 
-    def with(metadata_adapter:)
-      yield self.class.new(metadata_adapter: metadata_adapter, storage_adapter: storage_adapter)
+    # Given a resource in a change_set, recursively deletes all members for the entire
+    # tree of members.  Will delete all members first and then the resource itself
+    # to avoid issues with removing the root of the tree first.  However all other
```

Comment:
It's not a tree. It's a graph.  This method is extremely dangerous in such a situation.